### PR TITLE
Index stale operations to Lucene to have complete history

### PR DESCRIPTION
**Pull request by dnhatn**
Today, when processing out of order operations, we only add it into
translog but skip adding into Lucene. Translog, therefore, has a
complete history of sequence numbers while Lucene does not.

Since we would like to have a complete history in Lucene, this change
makes sure that stale operations will be added to Lucene as soft-deleted
documents if required.

### CODE REVIEW

**Comment by dnhatn** [comment]
@s1monw and @bleskes Can you please have a look? I could not make review requests. Thank you!

File: server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java [hunk 1]
```diff
@@ -193,4 +195,32 @@ public static long loadVersion(IndexReader reader, Term term) throws IOException
...
+        final List<LeafReaderContext> leaves = reader.leaves();
+        // iterate backwards to optimize for the frequently updated documents which are likely to be in the last segments
+        for (int i = leaves.size() - 1; i >= 0; i--) {
+            final LeafReaderContext leaf = leaves.get(i);
+            final PerThreadIDVersionAndSeqNoLookup lookup = lookups[leaf.ord];
+            final PostingsEnum postingsEnum = lookup.getPostingsOrNull(idTerm.bytes());
+            if (postingsEnum == null) {
+                continue;
+            }
+            final NumericDocValues seqNoDV = leaf.reader().getNumericDocValues(SeqNoFieldMapper.NAME);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java]: I think we can assert that both seqNoDV and primaryTermDV are non null. They are required at least for this code to work.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -577,23 +580,35 @@ public GetResult get(Get get, BiFunction<String, SearcherScope, Searcher> search
...
-        OP_STALE_OR_EQUAL,
+        /** the op is stale but its history is existed in Lucene */
+        OP_STALE_HISTORY_EXISTED,
+        /** the op is stale and its history is not found in Lucene */
+        OP_STALE_HISTORY_NOT_FOUND,
         /** no doc was found in lucene */
         LUCENE_DOC_NOT_FOUND
     }
 
+    private OpVsLuceneDocStatus compareToLuceneHistory(final Operation op, final Searcher searcher) throws IOException {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can we have a javadoc for this method?

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -577,23 +580,35 @@ public GetResult get(Get get, BiFunction<String, SearcherScope, Searcher> search
     enum OpVsLuceneDocStatus {
         /** the op is more recent than the one that last modified the doc found in lucene*/
         OP_NEWER,
-        /** the op is older or the same as the one that last modified the doc found in lucene*/
-        OP_STALE_OR_EQUAL,
+        /** the op is stale but its history is existed in Lucene */
+        OP_STALE_HISTORY_EXISTED,
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: maybe `OP_STALE_HISTORY_EXISTS` ?

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 4]
```diff
@@ -610,10 +625,10 @@ private OpVsLuceneDocStatus compareOpToLuceneDocBasedOnSeqNo(final Operation op)
                     if (op.primaryTerm() > existingTerm) {
                         status = OpVsLuceneDocStatus.OP_NEWER;
                     } else {
-                        status = OpVsLuceneDocStatus.OP_STALE_OR_EQUAL;
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can you please add comments here in what cases we can get into these situations?

**Review by s1monw — CHANGES_REQUESTED** [review]
I left some minor things. this looks pretty awesome! @bleskes please take a look at the engine parts.

**Comment by bleskes** [comment]
I talked to @dnhatn and I'm a bit uncomfortable with the PR changes to how we resolve stale operations in the engine. It seems that the main goal here was to try to avoid indexing duplicates of the same stale operation into Lucene. I'm not sure we need to do this. This is very rare. The alternative would be to accept duplicates under the condition that duplicate seq# also means the same doc version. A few things that are worth noting on top of it:
1) The Translog already has this property and may have duplicates.
2) To fully achieve these semantics we need to implement lucene roll backs.

If we accept the above, we can keep the current (simpler) way we resolve stale operations and follow the following:
1) If an incoming op with a seq# < local checkpoint, skip lucene and only store in the translog (as now)
2) If not, resolve the latest seq# for the doc id to decide if the op is stale (as today)
3) Stale ops with seq# > local checkpoint go into lucene with the soft delete flag "on".

**Comment by dnhatn** [comment]
>  The alternative would be to accept duplicates under the condition that duplicate seq# also means the same doc version.

@bleskes I pushed https://github.com/elastic/elasticsearch/pull/29679/commits/70d5359b871ca7f37bff01e885de8d6f78fed98f to back out the dedup logic as we discussed. Can you please have another look? Thank you!

File: server/src/main/java/org/elasticsearch/common/lucene/Lucene.java [hunk 5]
```diff
@@ -833,6 +834,58 @@ public int length() {
         };
     }
 
+    /**
+     * Wraps a directory reader to include all live docs.
+     * The wrapped reader can be used to query documents which are soft-deleted.
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/common/lucene/Lucene.java]: just say `to query all documents`

**Review by s1monw — APPROVED** [review]
left a tiny comment LGTM

**Comment by dnhatn** [comment]
@elasticmachine test this please.

**Comment by dnhatn** [comment]
run sample packaging tests

**Comment by dnhatn** [comment]
Thanks @simonw and @bleskes for reviewing.

**Pull request merged**